src/NHibernate/Transform/QueryAliasToObjectPropertySetter.cs

```diff
@@ -0,0 +1,197 @@
+		/// this instance has been built.</param>
+		/// <exception cref="PropertyNotFoundException">Thrown if no matching property or field can be found.</exception>
+		/// <exception cref="AmbiguousMatchException">Thrown if many matching properties or fields are found, having the
+		/// same visibility and inheritance depth.</exception>
+		public void SetProperty(string alias, object value, object resultObj)
+		{
+			if (_propertiesByNameCaseSensitive.TryGetValue(alias, out var property))
+			{
+				checkMember(property);
+				property.Member.SetValue(resultObj, value, new object[0]);
+				return;
+			}
+			if (_fieldsByNameCaseSensitive.TryGetValue(alias, out var field))
+			{
+				checkMember(field);
+				field.Member.SetValue(resultObj, value);
+				return;
+			}
+			if (_propertiesByNameCaseInsensitive.TryGetValue(alias, out property))
+			{
+				checkMember(property);
+				property.Member.SetValue(resultObj, value, new object[0]);
+				return;
+			}
+			if (_fieldsByNameCaseInsensitive.TryGetValue(alias, out field))
+			{
+				checkMember(field);
+				field.Member.SetValue(resultObj, value);
+				return;
+			}
+
+			throw new PropertyNotFoundException(resultObj.GetType(), alias, "setter");
+
+			void checkMember<T>(NamedMember<T> member) where T : MemberInfo
+			{
+				if (member.Member != null)
+					return;
+
+				if (member.AmbiguousMembers == null || member.AmbiguousMembers.Length < 2)
+					throw new InvalidOperationException($"{nameof(NamedMember<T>.Member)} missing and {nameof(NamedMember<T>.AmbiguousMembers)} invalid.");
+
+				throw new AmbiguousMatchException(
+					$"Unable to find adequate property or field to set on '{member.AmbiguousMembers[0].DeclaringType.Name}' for alias '{alias}', " +
+						$"many {(member.AmbiguousMembers[0] is PropertyInfo ? "properties" : "fields")} matches: {string.Join(", ", member.AmbiguousMembers.Select(m => m.Name))}");
+			}
+		}
+
+		/// <summary>
+		/// Build a <c>QueryAliasToObjectPropertySetter</c> for a supplied type.
+		/// </summary>
+		/// <param name="objType">The type.</param>
+		/// <returns>A <c>QueryAliasToObjectPropertySetter</c>.</returns>
+		public static QueryAliasToObjectPropertySetter MakeFor(System.Type objType)
+		{
+			var bindingFlags = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly;
+			var fields = new List<RankedMember<FieldInfo>>();
+			var properties = new List<RankedMember<PropertyInfo>>();
+			var currentType = objType;
+			var rank = 1;
+			// For grasping private members, we need to manually walk the hierarchy.
+			while (currentType != null && currentType != typeof(object))
+			{
+				fields.AddRange(
+					currentType
+						.GetFields(bindingFlags)
+						.Select(f => new RankedMember<FieldInfo> { Member = f, VisibilityRank = getFieldVisibilityRank(f), HierarchyRank = rank }));
+				properties.AddRange(
+					currentType
+						.GetProperties(bindingFlags)
+						.Where(p => p.CanWrite)
```

Comment:
Need to check [`p.GetIndexParameters().Length ==0`](https://msdn.microsoft.com/en-us/library/system.reflection.propertyinfo.getindexparameters.aspx)